engine/source/assets/ShaderAssetManager.cpp

```diff
@@ -0,0 +1,239 @@
+
+        if (descriptorType == nc::DescriptorType::None)
+        {
+            throw nc::NcError("No valid descriptor type found.");
+        }
+
+        std::string stagesFromFile;
+        file >> stagesFromFile;
+        auto stages = nc::ShaderStages::None;
+        if (stagesFromFile.find("Vertex") != std::string::npos) 
+        {
+            stages = stages | nc::ShaderStages::Vertex;
+        }
+        if (stagesFromFile.find("Fragment") != std::string::npos) 
+        {
+            stages = stages | nc::ShaderStages::Fragment;
+        }
+        if (stages == nc::ShaderStages::None)
+        {
+            throw nc::NcError("No valid shader stages found.");
+        }
+
+        return nc::DescriptorManifest
+        {
+            .setIndex = set,
+            .slotIndex = binding,
+            .descriptorType = descriptorType,
+            .shaderStages = stages
+        };
+    }
+
+    auto ReadShader(const std::string& path, const std::string uid) -> nc::ShaderView
+    {
+        const auto ncaString = (std::filesystem::path(path) / std::filesystem::path("shader.nca")).string();
+
+        if (!HasValidAssetExtension(ncaString)) throw nc::NcError("Invalid extension: " + ncaString);
+
+        std::ifstream file{ncaString};
+        if (!file.is_open()) throw nc::NcError("Could not open file: " + ncaString);
+
+        std::string vertexShaderPath;
+        file >> vertexShaderPath;
+        vertexShaderPath = (std::filesystem::path(path) / std::filesystem::path(vertexShaderPath)).string();
+
+        std::string fragmentShaderPath;
+        file >> fragmentShaderPath;
+        fragmentShaderPath = (std::filesystem::path(path) / std::filesystem::path(fragmentShaderPath)).string();
+
+        auto vertexByteCode = ReadSpirvIntoBytes(vertexShaderPath);
+        auto fragmentByteCode = ReadSpirvIntoBytes(fragmentShaderPath);
+
+        auto shaderView = nc::ShaderView
+        {
+            .uid ="",
+            .vertexByteCode = vertexByteCode,
+            .fragmentByteCode =  fragmentByteCode,
+            .descriptors = {}
+        };
+
+        uint32_t descriptorCount;
+        file >> descriptorCount;
+
+        for (uint32_t i = 0; i < descriptorCount; ++i)
+        {
+            shaderView.descriptors.emplace_back(ReadDescriptor(file));
+        }
+
+        shaderView.uid = uid;
+
+        return shaderView;
+    }
+
+}
+
+namespace nc
+{
+    ShaderAssetManager::ShaderAssetManager(const std::string& assetDirectory)
+        : m_shaderViews{},
+          m_assetDirectory{assetDirectory}
+    {
+    }
+
+    bool ShaderAssetManager::Load(const std::string& path, bool isExternal)
+    {
+        const auto fullPath = isExternal ? path : m_assetDirectory + path;
+
+        auto shaderView = ReadShader(fullPath, path);
+
+        if (IsLoaded(fullPath))
+        {
+            return false;
+        }
+
+        for (const auto& manifest : m_shaderViews)
```

Comment:
Maybe I'm missing the point here, but isn't this just checking `IsLoaded()` again?